entitlement/src/test/java/org/killbill/billing/entitlement/api/TestDefaultEntitlement.java

```diff
@@ -71,6 +71,98 @@ public void testCancelWithEntitlementDate() throws AccountApiException, Entitlem
+        assertListenerStatus();
+        final Entitlement entitlement = entitlementApi.getEntitlementForId(entitlementId, callContext);
+        assertEquals(entitlement.getState(), EntitlementState.ACTIVE);
+        assertEquals(entitlement.getSourceType(), EntitlementSourceType.NATIVE);
+
+        clock.addDays(5);
+
+        testListener.pushExpectedEvents(NextEvent.CANCEL, NextEvent.BLOCK);
+        final DateTime cancelDateTime = new DateTime(clock.getUTCNow());
+        entitlement.cancelEntitlementWithDate(cancelDateTime, cancelDateTime, Collections.emptyList(), callContext);
+        assertListenerStatus();
+
+        final Entitlement entitlement2 = entitlementApi.getEntitlementForId(entitlement.getId(), callContext);
+        assertEquals(entitlement2.getState(), EntitlementState.CANCELLED);
+        assertEquals(internalCallContext.toLocalDate(entitlement2.getEffectiveEndDate()), clock.getUTCToday());
+        assertEquals(entitlement2.getSourceType(), EntitlementSourceType.NATIVE);
+    }
+
+    @Test(groups = "slow")
+    public void testCancelWithEntitlementDateTimeInFuture() throws AccountApiException, EntitlementApiException {
+        final LocalDate initialDate = new LocalDate(2013, 8, 7);
+        clock.setDay(initialDate);
+
+        final Account account = createAccount(getAccountData(7));
+
+        final PlanPhaseSpecifier planPhaseSpecifier = new PlanPhaseSpecifier("Shotgun", BillingPeriod.MONTHLY, PriceListSet.DEFAULT_PRICELIST_NAME, null);
+
+        // Create entitlement and check each field
+        testListener.pushExpectedEvents(NextEvent.CREATE, NextEvent.BLOCK);
+        final UUID entitlementId = entitlementApi.createBaseEntitlement(account.getId(), new DefaultEntitlementSpecifier(planPhaseSpecifier), account.getExternalKey(), null, null, false, true, Collections.emptyList(), callContext);
+        assertListenerStatus();
+        final Entitlement entitlement = entitlementApi.getEntitlementForId(entitlementId, callContext);
+        assertEquals(entitlement.getState(), EntitlementState.ACTIVE);
+        assertEquals(entitlement.getSourceType(), EntitlementSourceType.NATIVE);
+
+        final DateTime cancelDateTime = new DateTime(clock.getUTCNow().plusDays(5));
+        entitlement.cancelEntitlementWithDate(cancelDateTime, cancelDateTime, Collections.emptyList(), callContext);
+        assertListenerStatus();
+
+        clock.addDays(5);
+        testListener.pushExpectedEvents(NextEvent.CANCEL, NextEvent.BLOCK);
+        final Entitlement entitlement2 = entitlementApi.getEntitlementForId(entitlement.getId(), callContext);
+        assertEquals(entitlement2.getState(), EntitlementState.CANCELLED);
+        assertEquals(internalCallContext.toLocalDate(entitlement2.getEffectiveEndDate()), clock.getUTCToday());
```

Comment:
Can we do better and check for the exact timestamp (DateTime) to verify the provided value was indeed honored?

Reply:
Also, can we check both billing and entitlement canceled date(time) ?

Reply:
> Can we do better and check for the exact timestamp (DateTime) to verify the provided value was indeed honored?

 Earlier we had decided to compare dates (and not DateTime) in the tests to prevent test flakiness.  Since we are cancelling with an exact `DateTime` in this case, I _think_ the tests would not be flaxy even if we compare `DateTime`. However, this would make the tests look different from all the other tests (that compare `LocalDate`).

Let me know what you think

Reply:
> Earlier we had decided to compare dates (and not DateTime) in the tests to prevent test flakiness. 

Yes, because we had existing tests that were *only* comparing (local) dates as our models/apis did not support DateTime

However, now we are introducing new apis to support DateTime, and so the new tests should verify the equality of the DateTime (not just the local dates).

Reply:
Agreed, yes, makes sense, I will change these tests to compare DateTime 👍

Reply:
Fixed, added check for billing and entitlement cancelled date time and changed test to compare DateTime